### DISCUSSION TITLE
Is there a modify_params_after_update?

### DISCUSSION THREAD

**Discussion by guruvamsi-policharla** [qucontrol/krotov#91] [Q&A] [answered]
Currently, there's a modify_params_after_iter option which allows injection of arbitrary code after each _iteration_ of Krotov. 
Is there by any chance an option to inject arbitrary code after each _update_ (time step)?

**Comment by goerz** [comment, marked as answer]
There is not, but you can "inject code" for most use cases due to the fact that the `propagate` routine that you pass to `optimize_pulses` is under your control. Specifically, you have direct read/write access to both the propagated state and the Hamiltonian/Liouvillian.

The two specific use cases that I've come across are:

* Write out observables during the forward/backward propagation.
* Adjust a "moving grid" Hamiltonian, as e.g. in [Fürst et al. New J. Phys. 16, 075007 (2014)](https://dx.doi.org/10.1088/1367-2630/16/7/075007)

Both of these are very straightforward to do inside a custom `propagate` routine.

You could conceivably extend the scope of what `propagate` has access to with some Python trickery ([closures](https://www.programiz.com/python-programming/closure)), e.g. to give it access to the objectives:

~~~ python
def make_propagator(objectives):

    def propagator(H, state, dt, c_ops=None, backwards=False, initialize=False):
        pass
        # do the propagation, and something involving objectives

    return propagator

# optimize_pulses(objectives,..., propagator=make_propagator(objectives), ...)
~~~

For anything beyond that, you can always create your own copy of the `optimize_pulses` routine and modify that to your needs. I'd have to see a very convincing use case that isn't covered by one of the two approaches I outlined above to consider adding further hooks to the implementation in the `krotov` package.
